README.md
<div align="center"><img src="https://user-images.githubusercontent.com/10853207/209234630-b29fbaaa-536b-4899-8eda-3a42a2d73023.png" width=300></div>

<p align="center"></p>

<div align="center">

![License](https://img.shields.io/badge/license-GPLv3-green)
![GitHub go.mod Go version](https://img.shields.io/github/go-mod/go-version/blue-devil/nexeruncator)
[![Go Report Card](https://goreportcard.com/badge/github.com/blue-devil/nexeruncator)](https://goreportcard.com/report/github.com/blue-devil/nexeruncator)

</div>

# neXeruncator

neXeruncator; extracts or inserts javascript source file from or into nexe-compiled binaries.

* The word `neXeruncator` is derived from `nexe` and `aberuncate`. Nothing special!

## Build

neXeruncator project created using Go version 1.19.
Building project is as simple as running `go build`

```bash
go build main.go
```

## Install

neXeruncator can work as a standalone executable, you do not need to install it.
Apply the build step.

## Usage

Running `nexeruncator` prints help. There are two commands `extract` and `insert`.

```txt
          __    __                              _
 _ __   __\ \  / /__ _ __ _   _ _ __   ___ __ _| |_ ___  _ __
| '_ \ / _ \ \/ / _ | '__| | | | '_ \ / __/ _' | __/ _ \| '__|
| | | |  __/>  <  __| |  | |_| | | | | (_| (_| | || (_) | |
|_| |_|\___/ /\ \___|_|   \__,_|_| |_|\___\__,_| \_\___/|_|
          /_/  \_\                  Blue DeviL \_____/  SCT

NAME:
   nexeruncator - aberuncator of nexe-compiled binaries

USAGE:
   nexeruncator [global options] command [command options] [arguments...]

VERSION:
   0.3.0

COMMANDS:
   help     displays help messages.
   extract  extracts javascript source
   insert   inserts javascript source

GLOBAL OPTIONS:
   --version, -v  print the version (default: false)
```

### extract

`extract` command simply extracts embedded javascript file from nexe-compiled
binary.

```txt
NAME:
   nexeruncator extract - extracts javascript source

USAGE:
   nexeruncator extract [command options]

DESCRIPTION:
   Extracts javascript source file from nexe-compiled binary

OPTIONS:
   --file value  nexe-compiled binary
   --help, -h    show help (default: false)
```

Below, we can see a sample command line for extracting:

```bash
nexeruncator extract --file nexefile.exe
```

### insert

`insert` command inserts a new javascript source into nexe-compiled binary and
created a new file named: `patched.exe`. `insert` command takes 2 arguments:
nexe-compiled binary and new javascript:

```bash
NAME:
   nexeruncator insert - inserts javascript source

USAGE:
   nexeruncator insert [command options]

DESCRIPTION:
   Inserts javascript source file into nexe-compiled binary

OPTIONS:
   --dest value  nexe-compiled binary
   --src value   javascript source
   --help, -h    show help (default: false)
```

Below, we can see a sample command line for inserting javascript source into
nexe-compiled binary:

```bash
nexeruncator insert --dest nexefile.exe --src myjsfile.js
```

## Notes

From a reverse engineer's point of view, the nexe-compiled application consists of 4 main parts:

* NodeJS part, which contains NodeJS runtime and big in size, ~55mb
* Initializer javascript code.
* User's javacript code.
* 32 byte footer part.

Initializer javascript part contains name of user's javascript source filename. Let's say we have a javascript file "awesome.js". That means "awesome.js" string occurs two times in initializer part.

This is how the nexe-compiled binary ends:

```txt
Offset(h) 00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F

035E3820        3C 6E 65 78 65 7E 7E 73 65 6E 74 69 6E 65    <nexe~~sentine
035E3830  6C 3E 00 00 00 00 00 FE CE 40 00 00 00 00 00 00  l>.....þÎ@......
035E3840  43 40                                            C@
```

The above sniplet is 32 bytes. The first 16 byte is the string literal `<nexe~~sentinel>`. The second 16 byte is actually comprise of 2 8 byte parts. And they are in little endian. Let's check [Nexe][gh-nexe]'s sources: `nexe/src
/
compiler.ts`

```ts
<snip>
    const code = this.code(),
        codeSize = Buffer.byteLength(code),
        lengths = Buffer.from(Array(16))

    lengths.writeDoubleLE(codeSize, 0)
    lengths.writeDoubleLE(this.bundle.size, 8)
    return new (MultiStream as any)([
        binary,
        toStream(code),
        this.bundle.toStream(),
        toStream(Buffer.concat([Buffer.from('<nexe~~sentinel>'), lengths])),
    ])
  }
}
```

The first 8 byte is `codeSize` which is what I called jsInit. It is a kind og initializer javascript code just before user's javascript code and finished with double semi-coloumns. And the second 8 byte is size of user's appended javascript source file.

## Todo

* [ ] Implement a info function
  * [ ] Collect info about nexe-compiled binary(js source name, file size)
  * [ ] Print this info
* [ ] Add to pkg.go.dev as a module
* [ ] Finish documentation
* [ ] Implement tests
* [ ] Add pre-built binaries
* [ ] Add testdata

## Resources

* [Github - Nexe][gh-nexe]

## License

This project is under GPLv3 license.

[gh-nexe]: https://github.com/nexe/nexe/


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.3.0] - 2022-12-30

### Changed

- FIX 1: ClearScreen func updated for a cross-platform version

## [0.2.0] - 2022-12-27

### Added

- urfave v2 implemented for commandline parameters
- commands added for extracting and inserting
- Added Todo
- Added `goreportcard`

### Changed

- not using flag package for commandline parameters
- not using import anymore. using insert

### Fixed

- Links which refers old repo are fixed

## [0.1.0] - 2022-12-23

Initial release.

nexeruncator can now extract javascript source from a nexe-compiled binary.
It exports and and saved the javascript source as its original name.
nexeruncator can also insert a new javascript source file into nexe-compiled
binary.

[0.3.0]: https://github.com/blue-devil/nexeruncator/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/blue-devil/nexeruncator/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/blue-devil/nexeruncator/releases/tag/v0.1.0


nexeruncator/commands/help/help_test.go
package help_test

import (
	"os"
	"testing"

	"github.com/blue-devil/nexeruncator/nexeruncator/commands/extractjs"
	"github.com/blue-devil/nexeruncator/nexeruncator/commands/help"

	"github.com/stretchr/testify/require"
	"github.com/urfave/cli/v2"
)

func TestHelp(t *testing.T) {
	exeName, err := os.Executable()
	require.NoError(t, err)

	app := &cli.App{
		Commands: []*cli.Command{
			help.Command(),
			extractjs.Command(),
		},
	}
	testArgs := []string{exeName, "help"}
	require.NoError(t, app.Run(testArgs))

	testArgs = []string{exeName, "help", "key"}
	require.NoError(t, app.Run(testArgs))

}


nexeruncator/commands/insertjs/insertjs_test.go
package insertjs_test


nexeruncator/commands/extractjs/extractjs_test.go
package extractjs_test


main.go
// nexeruncator - exports/inserts javascript from nexe-compiled binaries
// Copyright (C) 2022  BlueDeviL // SCT <bluedevil.SCT@gmail.com>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

package main

import (
	"io"
	"log"
	"os"

	"github.com/blue-devil/nexeruncator/nexeruncator/commands/extractjs"
	"github.com/blue-devil/nexeruncator/nexeruncator/commands/help"
	"github.com/blue-devil/nexeruncator/nexeruncator/commands/insertjs"
	"github.com/blue-devil/nexeruncator/nexeruncator/pkg/utils"

	"github.com/urfave/cli/v2"
)

var Version = "0.3.0"
var AppName = "nexeruncator"

func main() {
	utils.PrintBanner()

	app := &cli.App{
		Name:     AppName,
		Usage:    "aberuncator of nexe-compiled binaries",
		Commands: Commands(os.Stdin),
		Version:  Version,
	}

	if err := app.Run(os.Args); err != nil {
		log.Fatal(err)
	}
}

func Commands(reader io.Reader) []*cli.Command {
	return []*cli.Command{
		help.Command(),
		extractjs.Command(),
		insertjs.Command(),
	}
}


nexeruncator/nexeruncator.go
// Package nexeruncator provides two functionalities. nexeruncator can
// extract embedded javascript source from nexe-compiled binaries.
// nexeruncator can insert a new javascript source to a nexe-compiled
// binary
package nexeruncator

import (
	"log"

	"github.com/blue-devil/nexeruncator/nexeruncator/pkg/utils"
)

// ExtractJavascript function extracts javascript source file from nexeFile
func ExtractJavascript(nexeFile string) {
	var err error
	if err = utils.ExtractJS(nexeFile); err != nil {
		log.Printf("[-] Failed to extract javasript source file from %s, %v", nexeFile, err)
	}
}

// InsertJavascript function inserts jsFile into nexefile. Both nexeFile
// and jsFile is the name of files. Function did not overwrites the original
// executable file, creates a new file with name `pathced.exe`
func InsertJavascript(nexeFile, jsFile string) {
	var err error
	if err = utils.InsertJS(nexeFile, jsFile); err != nil {
		log.Printf("[-] Failed to insert javascript source file %s to %s, error: %v", jsFile, nexeFile, err)
	}
}


nexeruncator/commands/insertjs/insertjs.go
package insertjs

import (
	"log"

	"github.com/blue-devil/nexeruncator/nexeruncator/pkg/utils"

	"github.com/urfave/cli/v2"
)

const (
	CmdInsert = "insert"
	flagDest  = "dest"
	flagSrc   = "src"
)

func Command() *cli.Command {
	return &cli.Command{
		Name:        CmdInsert,
		HelpName:    CmdInsert,
		Action:      Action,
		ArgsUsage:   ` `,
		Usage:       `inserts javascript source`,
		Description: `Inserts javascript source file into nexe-compiled binary`,
		Flags:       Flags(),
	}
}

func Flags() []cli.Flag {
	return []cli.Flag{
		&cli.StringFlag{
			Name:     flagDest,
			Usage:    "nexe-compiled binary",
			Required: true,
		},
		&cli.StringFlag{
			Name:     flagSrc,
			Usage:    "javascript source",
			Required: true,
		},
	}
}

func Action(c *cli.Context) error {
	var nexeFile string
	var jsFile string
	var err error

	if c.IsSet(flagDest) && c.IsSet(flagSrc) {
		nexeFile = c.String(flagDest)
		jsFile = c.String(flagSrc)
	}
	if err = utils.InsertJS(nexeFile, jsFile); err != nil {
		log.Printf("[-] Failed to insert javascript source file %s into %s, error: %v", jsFile, nexeFile, err)
	}
	return nil
}


nexeruncator/commands/extractjs/extractjs.go
package extractjs

import (
	"log"

	"github.com/blue-devil/nexeruncator/nexeruncator/pkg/utils"

	"github.com/urfave/cli/v2"
)

const (
	CmdExtract = "extract"

	flagFile = "file"
)

func Command() *cli.Command {
	return &cli.Command{
		Name:        CmdExtract,
		HelpName:    CmdExtract,
		Action:      Action,
		ArgsUsage:   ` `,
		Usage:       `extracts javascript source`,
		Description: `Extracts javascript source file from nexe-compiled binary`,
		Flags:       Flags(),
	}
}

func Flags() []cli.Flag {
	return []cli.Flag{
		&cli.StringFlag{
			Name:     flagFile,
			Usage:    "nexe-compiled binary",
			Required: true,
		},
	}
}

func Action(c *cli.Context) error {
	var nexeFile string
	var err error

	if c.IsSet(flagFile) {
		nexeFile = c.String(flagFile)
	}

	if err = utils.ExtractJS(nexeFile); err != nil {
		log.Printf("[-] Failed to extract javasript source file from %s, %v", nexeFile, err)
	}

	return nil
}


nexeruncator/pkg/utils/utils.go
// Package utils contains main logic of extracting and inserting javascript
// file from nexe-compiled binaries
package utils

import (
	"bytes"
	"fmt"
	"math"
	"os"
	"os/exec"
	"runtime"
	"strconv"
)

var MagicBytes = []byte{
	0x21, 0x28, 0x66, 0x75, 0x6E, 0x63, 0x74, 0x69, 0x6F, 0x6E, 0x20, 0x28,
	0x29, 0x20, 0x7B, 0x70, 0x72, 0x6F, 0x63, 0x65, 0x73, 0x73, 0x2E, 0x5F,
	0x5F, 0x6E, 0x65, 0x78, 0x65, 0x20, 0x3D, 0x20, 0x7B, 0x22, 0x72, 0x65,
	0x73, 0x6F, 0x75, 0x72, 0x63, 0x65, 0x73, 0x22, 0x3A, 0x7B, 0x22, 0x2E,
	0x2F,
}

// Checks OS and slears command prompt/terminal
func ClearConsole() {
	ros := runtime.GOOS
	switch ros {
	case "windows":
		cmd := exec.Command("cmd", "/c", "cls")
		cmd.Stdout = os.Stdout
		cmd.Run()
	case "linux":
		cmd := exec.Command("clear") //Linux example, its tested
		cmd.Stdout = os.Stdout
		cmd.Run()
	case "darwin":
		cmd := exec.Command("clear") //Linux example, its tested
		cmd.Stdout = os.Stdout
		cmd.Run()
	default:
		fmt.Printf("[-] OS %s not defined", ros)
	}
}

// CheckNoColor returns true if NO_COLOR environmental variable is set
func CheckNoColor() bool {
	_, exist := os.LookupEnv("NO_COLOR")
	return exist
}

// Prints nexeruncator banner to command prompt/terminal
func PrintBanner() {
	banner := `
          __    __                              _             
 _ __   __\ \  / /__ _ __ _   _ _ __   ___ __ _| |_ ___  _ __ 
| '_ \ / _ \ \/ / _ | '__| | | | '_ \ / __/ _' | __/ _ \| '__|
| | | |  __/>  <  __| |  | |_| | | | | (_| (_| | || (_) | |   
|_| |_|\___/ /\ \___|_|   \__,_|_| |_|\___\__,_| \_\___/|_|   
          /_/  \_\                  Blue DeviL \_____/  SCT
`

	ClearConsole()
	if CheckNoColor() {
		fmt.Println(banner)
	} else {
		colorBanner := fmt.Sprintf("\x1b[%dm%s\x1b[0m", 32, banner)
		fmt.Println(colorBanner)
	}
}

// Returns javascript source file's name
func getFilename(slice []byte) string {
	// search for MagicBytes in slice
	index := bytes.Index(slice, MagicBytes)
	if index == -1 {
		return ""
	}
	// MagicBytes found, create new slice starting at index + len(MagicBytes)
	filenameSlice := slice[index+len(MagicBytes):]
	// search for double quote in filenameSlice
	index = bytes.IndexByte(filenameSlice, '"')
	if index == -1 {
		return ""
	}
	// double quote found, create new slice from start to index
	filenameSlice = filenameSlice[:index]
	// convert slice to string and return
	return string(filenameSlice)
}

// Searches for javascript source inside nexe-compiled binary and if found
// saves the file as its own name
func ExtractJS(filename string) error {
	// read file into nexeFile slice
	nexeFile, err := os.ReadFile(filename)
	if err != nil {
		fmt.Printf("[-] Error reading file: %s\n", err)
		os.Exit(1)
	}
	// search for MagicBytes in nexeFile
	index := bytes.Index(nexeFile, MagicBytes)
	if index == -1 {
		fmt.Println("[-] Are you sure if this is a nexe compiled binary?")
		os.Exit(1)
	}
	// MagicBytes found, store index in nexeStart
	nexeStart := index
	// create jsWhole slice from nexeStart to end of nexeFile
	jsWhole := nexeFile[nexeStart:]
	// get filename from jsWhole slice
	jsFilename := getFilename(jsWhole)
	if jsFilename == "" {
		fmt.Println("[-] Cannot get filename...")
		os.Exit(1)
	}
	// search for ";;" in jsWhole slice
	index = bytes.Index(jsWhole, []byte(";;"))
	if index == -1 {
		fmt.Println("[-] Error: cannot find ;;")
		os.Exit(1)
	}

	// ";;" found, create script2 slice from index + 2 to end of jsWhole
	script2 := jsWhole[index+2:]
	// search for <nexe~~sentinel> in script2 slice
	index = bytes.Index(script2, []byte("<nexe~~sentinel>"))
	if index == -1 {
		fmt.Println("[-] Cannot find <nexe~~sentinel>...")
		os.Exit(1)
	}
	// <nexe~~sentinel> found, create finalScript slice from start of script2 to index
	finalScript := script2[:index]
	// save finalScript slice as binary file with jsFilename
	err = os.WriteFile(jsFilename, finalScript, 0644)
	if err != nil {
		fmt.Printf("[-] Error saving file: %s\n", err)
		os.Exit(1)
	}
	fmt.Printf("[+] Extracted script saved as %s\n", jsFilename)

	return nil
}

// Sets javascript sources's name and its size inside nexe-compiled binary.
func setFilenameAndSize(slice []byte, fileName string, size uint64) ([]byte, uint64) {
	magicBytes := []byte{0x3A, 0x7B, 0x22, 0x2E, 0x2F}
	// search for magicBytes in slice
	index := bytes.Index(slice, magicBytes)
	if index == -1 {
		// magicBytes not found
		return nil, 0
	}
	// magicBytes found, create retVal slice starting at index + len(magicBytes)
	retVal := make([]byte, 16000)
	copy(retVal, slice[:index+len(magicBytes)])
	// convert fileName to slice and append to retVal
	retVal = append(retVal, []byte(fileName)...)
	// append [0x22, 0x3A, 0x5B, 0x30, 0x2C] to retVal
	retVal = append(retVal, []byte{0x22, 0x3A, 0x5B, 0x30, 0x2C}...)
	// convert size to string and then to slice, append to retVal
	sizeStr := strconv.FormatUint(uint64(size), 10)
	retVal = append(retVal, []byte(sizeStr)...)
	// search for [0x5D, 0x7D, 0x7D, 0x3B, 0x0A, 0x7D, 0x29] in slice
	index = bytes.Index(slice, []byte{0x5D, 0x7D, 0x7D, 0x3B, 0x0A, 0x7D, 0x29})
	if index == -1 {
		// [0x5D, 0x7D, 0x7D, 0x3B, 0x0A, 0x7D, 0x29] not found
		return nil, 0
	}

	// search 0x65, 0x78, 0x65, 0x63, 0x50, 0x61, 0x74, 0x68, 0x29, 0x2C, 0x22, 0x2E, 0x2F
	index2 := bytes.Index(slice, []byte{0x65, 0x78, 0x65, 0x63, 0x50, 0x61, 0x74, 0x68, 0x29, 0x2C, 0x22, 0x2E, 0x2F})
	if index2 == -1 {
		// [0x65, 0x78, 0x65, 0x63, 0x50, 0x61, 0x74, 0x68, 0x29, 0x2C, 0x22, 0x2E, 0x2F] not found
		return nil, 0
	}

	// [0x5D, 0x7D, 0x7D, 0x3B, 0x0A, 0x7D, 0x29] found, append from index to end of slice to retVal
	temp := slice[index:(index2 + 13)]
	retVal = append(retVal, temp...)
	// convert fileName to slice and append to retVal
	retVal = append(retVal, []byte(fileName)...)

	//  search 0x22, 0x29, 0x0A, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x70
	index2 = bytes.Index(slice, []byte{0x22, 0x29, 0x0A, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x70})
	if index2 == -1 {
		// [0x65, 0x78, 0x65, 0x63, 0x50, 0x61, 0x74, 0x68, 0x29, 0x2C, 0x22, 0x2E, 0x2F] not found
		return nil, 0
	}
	retVal = append(retVal, slice[index2:]...)

	// return retVal and its size
	return retVal, uint64(len(retVal))
}

// Inserts and overwrites a new javascript file into a nexe-compiled binary
func InsertJS(nexeFile string, scriptFile string) error {
	// read script and nexe files as slices of bytes
	jsscript, err := os.ReadFile(scriptFile)
	if err != nil {
		fmt.Printf("[-] Error reading script file: %s\n", err)
		os.Exit(1)
	}
	wholeProgram, err := os.ReadFile(nexeFile)
	if err != nil {
		fmt.Printf("[-] Error reading nexe file: %s\n", err)
		os.Exit(1)
	}
	// store script file size and name
	lenScript := uint64(len(jsscript))
	nameScript := scriptFile

	// create engine slice from start of wholeProgram to magicBytes
	index := bytes.Index(wholeProgram, MagicBytes)
	if index == -1 {
		fmt.Println("[-] Are you sure if this is a nexe compiled binary?")
		os.Exit(1)
	}
	engine := wholeProgram[:index]

	// create jsWhole slice from index to end of wholeProgram
	jsWhole := wholeProgram[index:]
	// search for ";;" in jsWhole
	index = bytes.Index(jsWhole, []byte(";;"))
	if index == -1 {
		fmt.Println("[-] Error: cannot find ;;")
		os.Exit(1)
	}
	// ";;" found, create jsInit slice from start of jsWhole to index + 2
	jsInit := jsWhole[:index+2]

	//
	// NOTE(bluedevil): Gather info about files

	// store size of jsInit
	// oldCodeSize := uint64(len(jsInit))
	// fmt.Println(oldCodeSize)

	// call setFilenameAndSize with jsInit, nameScript, and lenScript
	jsInitFin, codeSize := setFilenameAndSize(jsInit, nameScript, lenScript)

	// convert lenScript and codeSize to float64 and make 8-byte slices
	lenScriptF64 := math.Float64bits(float64(lenScript))
	lenScriptBytes := []byte{
		byte(lenScriptF64 >> 56),
		byte(lenScriptF64 >> 48),
		byte(lenScriptF64 >> 40),
		byte(lenScriptF64 >> 32),
		byte(lenScriptF64 >> 24),
		byte(lenScriptF64 >> 16),
		byte(lenScriptF64 >> 8),
		byte(lenScriptF64),
	}
	// reverse and append lenScriptBytes and codeSizeBytes to create footer
	codeSizeF64 := math.Float64bits(float64(codeSize))
	codeSizeBytes := []byte{
		byte(codeSizeF64 >> 56),
		byte(codeSizeF64 >> 48),
		byte(codeSizeF64 >> 40),
		byte(codeSizeF64 >> 32),
		byte(codeSizeF64 >> 24),
		byte(codeSizeF64 >> 16),
		byte(codeSizeF64 >> 8),
		byte(codeSizeF64),
	}
	for i, j := 0, len(lenScriptBytes)-1; i < j; i, j = i+1, j-1 {
		lenScriptBytes[i], lenScriptBytes[j] = lenScriptBytes[j], lenScriptBytes[i]
	}
	for i, j := 0, len(codeSizeBytes)-1; i < j; i, j = i+1, j-1 {
		codeSizeBytes[i], codeSizeBytes[j] = codeSizeBytes[j], codeSizeBytes[i]
	}
	footer := []byte("<nexe~~sentinel>")
	footer = append(footer, codeSizeBytes...)
	footer = append(footer, lenScriptBytes...)

	// create patched slice by concatenating engine, jsInitFin, jsscript, and footer
	patched := append(engine, jsInitFin...)
	patched = append(patched, jsscript...)
	patched = append(patched, footer...)

	// write patched slice to patched.exe file
	err = os.WriteFile("patched.exe", patched, 0644)
	if err != nil {
		fmt.Printf("[-] Error writing patched file: %s\n", err)
		os.Exit(1)
	}

	fmt.Printf("[+] File %s inserted into %s\n", scriptFile, nexeFile)

	return nil
}
